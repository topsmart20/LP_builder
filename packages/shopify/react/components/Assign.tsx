import React from 'react';
import { Builder, BuilderStore } from '@builder.io/react';

interface AssignBlockProps {
  expression?: string;
  builderState?: BuilderStore;
}

export class AssignBlock extends React.Component<AssignBlockProps> {
  constructor(props: AssignBlockProps) {
    super(props);

    const { expression, builderState } = props;

    if (expression && builderState) {
      builderState.state.shopify?.liquid?.assign(expression);
    }
  }
}

Builder.registerComponent(AssignBlock, {
  name: 'Shopify:Assign',
  hideFromInsertMenu: true,
  inputs: [
    {
      name: 'expression',
      type: 'javascript',
      // TODO: type: 'code', language: 'javascript'
    },
  ],
});
