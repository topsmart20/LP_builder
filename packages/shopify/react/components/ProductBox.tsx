/** @jsx jsx */
import { jsx, css, keyframes } from '@emotion/core';

import React, { useEffect, useState } from 'react';
import { Builder, BuilderElement } from '@builder.io/sdk';
import { BuilderStore, BuilderStoreContext, BuilderBlocks } from '@builder.io/react';
import { modifyProduct } from '../functions/modify-product';

interface ProductBoxProps {
  product?: string | number;
  /** @todo implement this */
  variant?: string | number;
  builderState?: BuilderStore;
  builderBlock?: BuilderElement;
}

const spin = keyframes`
  0% {
    transform:rotate(0)
  }
  100% {
    transform:rotate(360deg)
  }
`;

function LoadingSpinner() {
  return (
    <div
      css={css`
        width: 4em;
        height: 4em;
        border-radius: 50%;
        position: relative;
        margin: 6rem auto;
        font-size: 1rem;
        text-indent: -9999em;
        border-top: 0.2em solid rgba(131, 132, 137, 0.2);
        border-right: 0.2em solid rgba(131, 132, 137, 0.2);
        border-bottom: 0.2em solid rgba(131, 132, 137, 0.2);
        border-left: 0.2em solid #454749;
        transform: translateZ(0);
        animation: 1.1s linear infinite ${spin};
      `}
    />
  );
}

export function ProductBox(props: ProductBoxProps) {
  const [productInfo, setProductInfo] = useState(null);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    updateProduct();
  }, [props.product]);

  function updateProduct() {
    setLoading(true);
    const { context } = props.builderState!;
    let productId = props.product || '';

    if (productId) {
      fetch(`https://builder.io/api/v1/shopify/products/${productId}.json?apiKey=${context.apiKey}`)
        .then(res => res.json())
        .then(data => {
          const { product } = data;
          const modifiedProduct = modifyProduct(product);
          setProductInfo(modifiedProduct);
          setLoading(false);
        })
        .catch(err => {
          console.error('Error fetching Shopify product', err);
          setLoading(false);
        });
    } else {
      setProductInfo(null);
    }
  }

  return loading && !productInfo ? (
    <LoadingSpinner />
  ) : (
    <BuilderStoreContext.Consumer>
      {store => (
        <BuilderStoreContext.Provider
          value={{
            ...store,
            state: {
              ...store.state,
              // TODO: communicate up this up to DataEditor to populate bindings dropdowns (how?)
              productInfo,
            },
          }}
        >
          <BuilderBlocks
            child
            parentElementId={props.builderBlock?.id}
            blocks={props.builderBlock?.children}
          />
        </BuilderStoreContext.Provider>
      )}
    </BuilderStoreContext.Consumer>
  );
}

Builder.registerComponent(ProductBox, {
  name: 'Shopify:ProductBox',
  canHaveChildren: true,
  hideFromInsertMenu: true,
  inputs: [
    {
      name: 'product',
      type: 'ShopifyProduct',
    },
  ],
});
