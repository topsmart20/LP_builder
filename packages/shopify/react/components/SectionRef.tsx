import React, { useState, useEffect } from 'react';
import { Builder } from '@builder.io/react';

interface SectionRefProps {
  class?: string;
  section?: string;
}
const cache: { [key: string]: string } = {};

export const SectionRef = (props: SectionRefProps) => {
  const [html, setHtml] = useState<string | null>(null);
  const sectionName = props.section?.split('/')[1].split('.')[0]!;

  useEffect(() => {
    if (cache[sectionName]) {
      return;
    }
    // Convert `sections/foo.liquid` -> `foo`
    if (!sectionName) {
      return;
    }
    if (cache[sectionName]) {
      setHtml(cache[sectionName]);
      return;
    }
    // This may look strange, but it is how Shopify's section rendering API
    // works https://shopify.dev/docs/themes/sections/section-rendering-api
    fetch(`${location.search ? '&' : '?'}section_id=${sectionName}`)
      .then(res => res.text())
      .then(text => {
        cache[sectionName] = text;
        setHtml(text);
      });
  }, [sectionName]);

  return (
    <div
      className={`shopify-section ${props.class}`}
      dangerouslySetInnerHTML={{ __html: html || cache[sectionName] || '' }}
    />
  );
};

Builder.registerComponent(SectionRef, {
  name: 'Shopify:SectionRef',
  hideFromInsertMenu: true,
});
