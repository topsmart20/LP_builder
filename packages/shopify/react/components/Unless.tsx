import * as React from 'react';
import builder, {
  Builder,
  BuilderStore,
  onChange,
  withBuilder,
  BuilderElement,
  stringToFunction,
  BuilderBlocks,
} from '@builder.io/react';

interface UnlessBlockProps {
  builderState?: BuilderStore;
  builderBlock?: BuilderElement;
  unlessBlocks?: BuilderElement[];
  elseBlocks?: BuilderElement[];
  expression?: string;
}

export class UnlessBlock extends React.Component<UnlessBlockProps> {
  private getMatchingBlocks() {
    const { expression, unlessBlocks, elseBlocks } = this.props;
    if (!expression) {
      return unlessBlocks;
    }
    const fn = stringToFunction(expression);
    const result = fn(
      this.props.builderState && this.props.builderState.state,
      null,
      this.props.builderBlock,
      builder,
      null,
      this.props.builderState!.update,
      Builder,
      this.props.builderState?.context
    );
    return result ? elseBlocks : unlessBlocks;
  }

  render() {
    const { unlessBlocks } = this.props;
    const blocks = this.getMatchingBlocks();
    if (!blocks) {
      return null;
    }
    const isUnless = blocks === unlessBlocks;

    return (
      <BuilderBlocks
        child
        parentElementId={this.props.builderBlock && this.props.builderBlock.id}
        blocks={blocks}
        dataPath={`component.options.${isUnless ? 'unlessBlocks' : 'elseBlocks'}`}
      />
    );
  }
}

const defaultBlock = {
  '@type': '@builder.io/sdk:Element',
  responsiveStyles: {
    large: {
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'stretch',
      flexShrink: '0',
      position: 'relative',
      marginTop: '30px',
      textAlign: 'center',
      lineHeight: 'normal',
      height: 'auto',
    },
  },
  component: {
    name: 'Text',
    options: {
      text: '<p>Enter some text...</p>',
    },
  },
};

withBuilder(UnlessBlock, {
  name: 'Shopify:Unless',
  hideFromInsertMenu: true,
  noWrap: true,
  inputs: [
    {
      name: 'expression',
      type: 'javascript',
    },
    {
      name: 'unlessBlocks',
      type: 'array',
      hideFromUI: true,
      defaultValue: [defaultBlock],
    },
    {
      name: 'elseBlocks',
      type: 'array',
      hideFromUI: true,
      defaultValue: [defaultBlock],
    },
  ],
});
