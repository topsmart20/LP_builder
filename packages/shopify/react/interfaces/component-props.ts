import { BuilderStore, BuilderElement } from '@builder.io/react';

export interface AssignBlockProps {
  expression?: string;
  builderState?: BuilderStore;
}

export interface CaptureBlockProps {
  expression?: string;
  variableName?: string;
  builderState?: BuilderStore;
}

export interface Branch {
  expression?: string;
  blocks: BuilderElement[];
}

export interface ConditionBlockProps {
  builderState?: BuilderStore;
  builderBlock?: BuilderElement;
  branches: Branch[];
}

export interface UnlessBlockProps {
  builderState?: BuilderStore;
  builderBlock?: BuilderElement;
  unlessBlocks?: BuilderElement[];
  elseBlocks?: BuilderElement[];
  expression?: string;
}

export interface StateProviderProps {
  builderBlock?: BuilderElement;
  state: any;
  context?: any;
}
