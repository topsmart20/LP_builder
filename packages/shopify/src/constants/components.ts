import { BuilderElement } from "@builder.io/sdk";
import { Options } from "../interfaces/options";

export interface ComponentInfo {
  component: (block: BuilderElement, options: Options, attributes?: string) => string;
  noWrap?: boolean;
  name: string;
}

export function component(info: ComponentInfo) {
  components[info.name] = info;
  return info.component;
}

export function getComponentInfo(name: string) {
  return components[name];
}

export function getComponent(name: string) {
  const info = getComponentInfo(name)
  return info && info.component;
}

export const components: { [key: string]: ComponentInfo | undefined } = {};
