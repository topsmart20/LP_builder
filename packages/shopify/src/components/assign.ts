import { BuilderElement } from '@builder.io/sdk';
import { Options } from '../interfaces/options';
import { component } from '../constants/components';

export const Assign = component({
  name: 'Shopify:Assign',
  noWrap: true,
  component: (block: BuilderElement, renderOptions: Options) => {
    const { options } = block.component!;
    const content = options.expression;

    return `{% assign ${content} %}`;
  },
});
