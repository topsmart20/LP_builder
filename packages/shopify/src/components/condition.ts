import { BuilderElement } from '@builder.io/sdk';
import { Options } from '../interfaces/options';
import { component } from '../constants/components';
import { ConditionBlockProps } from '../../react/interfaces/component-props';
import { blockToLiquid, getLiquidConditionExpresion } from '../functions/block-to-liquid';

export const Condition = component({
  name: 'Shopify:Condition',
  noWrap: true,
  component: (block: BuilderElement, renderOptions: Options) => {
    const options = block.component!.options as ConditionBlockProps;
    let str = '';
    for (let i = 0; i < options.branches.length; i++) {
      const branch = options.branches[i];
      const expression = branch.expression && getLiquidConditionExpresion(branch.expression);
      if (i === 0) {
        str += `{% if ${expression} %}`;
        str += branch.blocks.map(item => blockToLiquid(item, renderOptions)).join('\n');
      } else {
        if (expression) {
          str += `{% elsif ${expression} %}`;
        } else {
          str += '{% else %}';
        }
        str += branch.blocks.map(item => blockToLiquid(item, renderOptions)).join('\n');
      }
    }
    str += '{% endif %}';

    return str;
  },
});
