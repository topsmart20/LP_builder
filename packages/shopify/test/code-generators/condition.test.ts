import { Condition } from '../../src/components/condition';
import { el } from '../modules/helpers';
import { ConditionBlockProps } from '../../react/interfaces/component-props';

/**
 * Helper to turn ConditionBlockProps into liquid for testing the output
 *
 * @example
 *    getConditionCode({ branches: [] })
 *
 * @returns The liquid code
 */
const getConditionCode = (options: ConditionBlockProps): string =>
  Condition(
    el({
      component: {
        options,
        name: 'Shopify:Condition',
      },
    }),
    {}
  );

// Helpers
const blockA = el({ id: 'blockA' });
const blockB = el({ id: 'blockB' });
const blockC = el({ id: 'blockC' });
const blockD = el({ id: 'blockD' });

const liquidExpressionA = 'products | first';
const liquidExpressionB = 'products | second';
const liquidExpressionC = 'products | third';

describe('Condition code generator', () => {
  test('Simple test', () => {
    const code = getConditionCode({
      branches: [
        {
          expression: `context.shopify.liquid.condition('${liquidExpressionA}', state)`,
          blocks: [blockA],
        },
      ],
    });

    expect(code).toMatchSnapshot();
  });
  test('2 branches', () => {
    const code = getConditionCode({
      branches: [
        {
          expression: `context.shopify.liquid.condition('${liquidExpressionA}', state)`,
          blocks: [blockA],
        },
        {
          expression: `context.shopify.liquid.condition('${liquidExpressionB}', state)`,
          blocks: [blockB],
        },
      ],
    });

    expect(code).toMatchSnapshot();
  });

  test('4 branches', () => {
    const code = getConditionCode({
      branches: [
        {
          expression: `context.shopify.liquid.condition('${liquidExpressionA}', state)`,
          blocks: [blockA],
        },
        {
          expression: `context.shopify.liquid.condition('${liquidExpressionB}', state)`,
          blocks: [blockB],
        },
        {
          expression: `context.shopify.liquid.condition('${liquidExpressionC}', state)`,
          blocks: [blockC],
        },
        {
          blocks: [blockD],
        },
      ],
    });

    expect(code).toMatchSnapshot();
  });
});
