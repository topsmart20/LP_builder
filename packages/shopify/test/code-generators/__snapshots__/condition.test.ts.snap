// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Condition code generator 2 branches 1`] = `
"{% if products | first %}
    <style>.builder-block.blockA {}</style>
    
    
    
    
      <div  builder-id=\\"blockA\\" class=\\"builder-block blockA\\">
        
      </div>
    
    
    
  {% elsif products | second %}
    <style>.builder-block.blockB {}</style>
    
    
    
    
      <div  builder-id=\\"blockB\\" class=\\"builder-block blockB\\">
        
      </div>
    
    
    
  {% endif %}"
`;

exports[`Condition code generator 4 branches 1`] = `
"{% if products | first %}
    <style>.builder-block.blockA {}</style>
    
    
    
    
      <div  builder-id=\\"blockA\\" class=\\"builder-block blockA\\">
        
      </div>
    
    
    
  {% elsif products | second %}
    <style>.builder-block.blockB {}</style>
    
    
    
    
      <div  builder-id=\\"blockB\\" class=\\"builder-block blockB\\">
        
      </div>
    
    
    
  {% elsif products | third %}
    <style>.builder-block.blockC {}</style>
    
    
    
    
      <div  builder-id=\\"blockC\\" class=\\"builder-block blockC\\">
        
      </div>
    
    
    
  {% else %}
    <style>.builder-block.blockD {}</style>
    
    
    
    
      <div  builder-id=\\"blockD\\" class=\\"builder-block blockD\\">
        
      </div>
    
    
    
  {% endif %}"
`;

exports[`Condition code generator Simple test 1`] = `
"{% if products | first %}
    <style>.builder-block.blockA {}</style>
    
    
    
    
      <div  builder-id=\\"blockA\\" class=\\"builder-block blockA\\">
        
      </div>
    
    
    
  {% endif %}"
`;
